rotkehlchen/chain/ethereum/modules/lido/decoder.py

```diff
@@ -0,0 +1,526 @@
+
+            if sth_wrapped_amount and wsteth_received_amount:
+                paired_wsteth_ratio = FVal(sth_wrapped_amount) / FVal(wsteth_received_amount)
+                break
+
+        if paired_wsteth_ratio is None:
+            return DEFAULT_DECODING_OUTPUT
+
+        in_event = None
+        out_event = None
+        for event in context.decoded_events:
+            if (
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.asset == A_STETH and
+                event.balance.amount == sth_wrapped_amount and
+                event.address == wsteth_evm_address
+            ):
+                event.event_type = HistoryEventType.DEPOSIT
+                event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                event.counterparty = CPT_LIDO_ETH
+                event.notes = f'Sent {sth_wrapped_amount} {A_STETH.symbol_or_name()} for wrapping at wstETH ratio {paired_wsteth_ratio}'  # noqa: E501
+                out_event = event
+            elif (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.asset == A_WSTETH and
+                event.balance.amount == wsteth_received_amount and
+                event.address == ZERO_ADDRESS
+            ):
+                event.event_type = HistoryEventType.RECEIVE
+                event.event_subtype = HistoryEventSubType.RECEIVE_WRAPPED
+                event.counterparty = CPT_LIDO_ETH
+                event.notes = f'Receive {wsteth_received_amount} {A_WSTETH.symbol_or_name()} in exchange of the deposited {A_STETH.symbol_or_name()} at wstETH ratio {paired_wsteth_ratio}'  # {amount} to be replaced in post decoding  # noqa: E501,
+                in_event = event
+        if in_event and out_event:
+            maybe_reshuffle_events(
+                ordered_events=[out_event, in_event],
+                events_list=context.decoded_events,
+            )
+
+        return DecodingOutput(matched_counterparty=CPT_LIDO_ETH, refresh_balances=True)
+
+    def _decode_lido_transfer_shares_for_wsteth_unwrap(
+            self, context: DecoderContext,
+    ) -> DecodingOutput:
+        """Decode wsteth unwrapping in steth from TransferShares log event"""
+
+        user_address = hex_or_bytes_to_address(context.tx_log.topics[2])
+
+        paired_wsteth_ratio = None
+        # we need to retrieve the transferred amount deterministically.
+        # log index baseline: transferShares log index
+        # expect log index for the transferred steth is -1
+        # expect log index for the transferred steth is -2
+        expected_steth_transfer_index = context.tx_log.log_index - 1
+        expected_wsteth_transfer_index = context.tx_log.log_index - 2
+        sth_received_amount = None
+        wsteth_unwrapped_amount = None
+        for tx_log in context.all_logs:
```

Comment:
again iterating all log events of the sender. This is not how decoders work. This is extremely slow